wagon-providers/wagon-http-shared/src/main/java/org/apache/maven/wagon/shared/http/AbstractHttpClientWagon.java

```diff
@@ -279,6 +332,20 @@ public boolean isStreaming()
     private static final boolean SSL_ALLOW_ALL =
         Boolean.valueOf( System.getProperty( "maven.wagon.http.ssl.allowall", "false" ) );
 
+    /**
+     * If enabled, then the content-type HTTP header will be set using the file extension
+     * or the stream header to determine the type, <b>enabled by default</b>
+     */
+    private static final boolean AUTOSET_CONTENT_TYPE =
+            Boolean.valueOf( System.getProperty( "maven.wagon.http.autocontenttype", "true" ) );
+
+    /**
+     * If enabled, then an when determining the content type will result in a fatal exception
+     * <b>disabled by default</b>
+     * This flag is only effective when maven.wagon.http.autocontenttype is set.
+     */
+    private static final boolean AUTOSET_CONTENT_TYPE_FATAL =
```

Comment:
YAGNI. No one needs this. Failure should unconditionally set the content-type to application/octet-stream.

Reply:
@elharo "Failure should unconditionally set the content-type to application/octet-stream." that is a slight difference from current functionality which does not set the content-type. Just to confirm the resulting behavior is what you are asking for: 
if the autoset content type is enabled then a content-type will always be added and the default is application/octet-stream. 
If autoset content type is not enabled then the content-type is not set. (current functionality)

Reply:
Maintaining current behavior might be important. I'm OK with not setting this is that's what this if it's necessary to maintain compatibility (though won't this detect and set in cases where it's not detected and set now?) But please don't throw a fatal exception here. This is a recoverable case with reasonable defaults.

Reply:
I think that the following behavior is simple to explain and therefore favored:
1.) if autoset is enabled then a content-type header will always be written and defaults to application/octet-stream when the type cannot be determined
2.) if autoset is not enabled then the content-type is not written (unless it is done externally to the wagon library)

So, no exception and the second flag to control the exception behavior is not needed.
How does that sound?
Finally, should autoset behavior be enabled or disabled by default? disabled maintains current behavior, enabled could be said to be consistent with the content-length header, which is always provided.